plat/kvm/x86/smp.c

```diff
@@ -0,0 +1,539 @@
+	uk_pr_debug("Copied AP 16 bit boot code to 0x8000\n");
+
+	/* Store the CR3 register */
+	__asm__ __volatile__("movq %%cr3, %0" : "=r"(cr3_val));
+
+	ret = ukplat_irq_register(UKPLAT_X86_IPI_RUN_IRQ,
+				  x86_lcpu_ipi_run_handler, NULL);
+	if (ret != 0) {
+		uk_pr_crit("Interrupt handler for IRQ %d failed\n",
+			   UKPLAT_X86_IPI_RUN_IRQ);
+		return ret;
+	}
+
+	for (i = 0; i < num; i++) {
+		id = (lcpuid) ? lcpuid[i] : (__lcpuid)i;
+
+		/* Don't restart the BootStrapping Processor */
+		if (id == bspid)
+			continue;
+
+		/*
+		 * Unitialized apic_id; either it wasn't found, or we don't
+		 * want to start it
+		 */
+		if (cpus[id].apic_id == 0)
+			continue;
+
+		UK_ASSERT(sp[i]);
+
+		/* Store the entry address and stack pointer */
+		cpus[id].s_args.entry = entry[i];
+		cpus[id].s_args.stackp = sp[i];
+
+		/* Store the interrupt stacks */
+		cpus[id].intr_stack = cpu_intr_stack[id];
+		cpus[id].trap_stack = cpu_trap_stack[id];
+		cpus[id].panic_stack = cpu_nmi_stack[id];
+
+		/* Initialize the lists */
+		UK_INIT_LIST_HEAD(&cpus[id].fnlist);
+		UK_INIT_LIST_HEAD(&cpus[id].nmi_fnlist);
+
+		/* clear APIC errors */
+		x2apic_clear_errors();
+
+		/* select AP and trigger INIT IPI */
+		x2apic_send_iipi(cpus[id].apic_id);
+
+		/* deassert */
+		x2apic_send_iipi_deassert(cpus[id].apic_id);
+	}
+
+	/* wait 10 msec */
+	mdelay(10);
+
+	for (i = 0; i < num; i++) {
+		id = (lcpuid) ? lcpuid[i] : (__lcpuid)i;
+
+		if (id == bspid)
+			continue;
+
+		if (cpus[id].apic_id == 0)
+			continue;
+
+		for (j = 0; j < 2; j++) {
+			/* clear APIC errors */
+			x2apic_clear_errors();
+
+			/**
+			 * select AP and trigger STARTUP IPI with entry code
+			 * at 0x8000
+			 */
+			x2apic_send_sipi(0x8000, cpus[id].apic_id);
+
+			/* wait 200 usec */
+			udelay(200);
+		}
+	}
+	/* wait 10 msec */
+	mdelay(10);
+
+	/* Signal the APs that the BSP has finished */
+	/* NOTE: without scheduling implemented, everything works fine without
+	 * this part (from here, until return), but everyone recommends that the
+	 * APs wait for the BSP to finish.
+	 */
+	bspdone = 1;
+
+	return 0;
+}
+
+__lcpuid ukplat_lcpu_id(void)
+{
+	__u32 eax, ebx, ecx, edx;
+
+	/*
+	 * Until we implement per-CPU storage, we must use cpuid to find out
+	 * the id of the current CPU
+	 */
+	cpuid(1, 0, &eax, &ebx, &ecx, &edx);
+
+	return (ebx >> 24);
+}
+
+int __attribute__((optimize("O0")))
+ukplat_lcpu_wait(__lcpuid lcpuid[], unsigned int num, __nsec timeout)
+{
+	unsigned int i, ncores = ukplat_lcpu_count();
+	__lcpuid id;
+	__nsec end;
+
+	num = (lcpuid) ? num : ncores;
+	if (num > ncores) {
+		uk_pr_warn("More cores are selected, but not available\n");
+		num = ncores;
+	}
+
+	if (timeout > 0)
+		end = ukplat_monotonic_clock() + timeout;
+
+	for (i = 0; i < num; i++) {
+		id = (lcpuid) ? lcpuid[i] : (__lcpuid)i;
+
+		/* Don't wait for the current CPU */
+		if (id == ukplat_lcpu_id())
+			continue;
+
+		while (cpus[id].state != LCPU_STATE_IDLE) {
```

Comment:
@cristian-vijelie I had a synchronization issue with this function in the ARM64 implementation. There is a situation when a cpu makes the following calls:
 ```
 ukplat_lcpu_run(target_cpu....);
 ukplat_lcpu_wait(target_cpu....);
 ukplat_lcpu_run(target_cpu....);
 ```
 Now, there is a scenario when the source cpu can reach this state check before the target one updated the state to `LCPU_STATE_BUSY`. Therefore, the source will exit `ukplat_lcpu_wait()` before the target core finished it's job, will send the `LCPU_RUN` interrupt while the target core has interrupts still disabled, so the second `ukplat_lcpu_run()` call will be lost.
 I have solved this using a lock associated to each cpu [1], here is a figure with the race condition:
![Uploading image.png…]()
[1] https://github.com/unikraft/unikraft/pull/373

Reply:
Hey @razvanvirtan,

thanks for pointing out the issue and for the detailed explanation. Also for you @cristian-vijelie, I am currently working on this PR internally and will address all of these issues. I am also working on merging initialization code for BSP and AP and also fix some issues concerning missing initialization of CPU features (e.g., at the moment the CPU caches are disabled for APs). @cristian-vijelie Please do not push any updates until further notice.

@razvanvirtan I am also looking at the ARM version. When I finished my first iteration on the x86 SMP code, I will take a look at the ARM one. Afterwards, we can see if there is some more stuff we can define/move/merge in a platform/architecture agnostic way to avoid duplicate code.